File: thousands-services/IvsIdleGameShared/Models/Wildcard/MatchResult.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;
using System;
using System.Collections.Generic;

namespace IvsIdleGameShared.Models.Wildcard;

[BsonIgnoreExtraElements]
public class MatchResult
{
    [BsonId]
    public ObjectId Id { get; set; }

    [BsonElement("matchResults")]
    public MatchResults MatchResults { get; set; }

    [BsonElement("createdAt")]
    public DateTime CreatedAt { get; set; }
}

[BsonIgnoreExtraElements]
public class MatchResults
{
    [BsonElement("lobbyId")]
    public string LobbyId { get; set; }

    [BsonElement("playerData")]
    public List<PlayerData> PlayerData { get; set; }

    [BsonElement("winningTeamId")]
    public long WinningTeamId { get; set; }

    [BsonElement("duration")]
    public double Duration { get; set; }

    [BsonElement("eventId")]
    public string EventId { get; set; }

    [BsonElement("matchId")]
    public string MatchId { get; set; }

    [BsonElement("gameTypeId")]
    public string GameTypeId { get; set; }

    [BsonElement("createdAt")]
    public DateTime CreatedAt { get; set; }
}

[BsonIgnoreExtraElements]
public class PlayerData
{
    [BsonElement("gamerTag")]
    public long GamerTag { get; set; }

    [BsonElement("teamId")]
    public int TeamId { get; set; }

    [BsonElement("loadout")]
    public Loadout Loadout { get; set; }

    [BsonElement("castCards")]
    public List<CastCard> CastCards { get; set; }

    [BsonElement("kos")]
    public int Kos { get; set; }

    [BsonElement("totalDamageDone")]
    public int TotalDamageDone { get; set; }

    [BsonElement("totalDamageReceived")]
    public int TotalDamageReceived { get; set; }

    [BsonElement("totalHealingInflicted")]
    public int TotalHealingInflicted { get; set; }

    [BsonElement("totalHealingReceived")]
    public int TotalHealingReceived { get; set; }

    [BsonElement("knockedouts")]
    public int Knockedouts { get; set; }

    [BsonElement("summonKOs")]
    public int SummonKOs { get; set; }

    [BsonElement("goalieKOs")]
    public int GoalieKOs { get; set; }

    [BsonElement("sidekicksSpawned")]
    public int SidekicksSpawned { get; set; }

    [BsonElement("healthCollected")]
    public long HealthCollected { get; set; }

    [BsonElement("manaCollected")]
    public long ManaCollected { get; set; }

    [BsonElement("damageTracking")]
    public Dictionary<string, long> DamageTracking { get; set; }

    [BsonElement("statusTracking")]
    public Dictionary<string, long> StatusTracking { get; set; }

    [BsonElement("totalWildcardsAcquired")]
    public long TotalWildcardsAcquired { get; set; }
}

[BsonIgnoreExtraElements]
public class Loadout
{
    [BsonElement("_id")]
    public string Id { get; set; }

    [BsonElement("championId")]
    public string ChampionId { get; set; }

    [BsonElement("name")]
    public string Name { get; set; }

    [BsonElement("summonCardIds")]
    public List<string> SummonCardIds { get; set; }

    [BsonElement("talentCardIds")]
    public List<string> TalentCardIds { get; set; }

    [BsonElement("wildCardIds")]
    public List<string> WildCardIds { get; set; }

    [BsonElement("cosmeticCardIds")]
    public List<string> CosmeticCardIds { get; set; }

    [BsonElement("equippedCosmetics")]
    public Dictionary<string, object> EquippedCosmetics { get; set; }

    [BsonElement("isDefault")] public bool? IsDefault { get; set; } = false;
}

[BsonIgnoreExtraElements]
public class CastCard
{
    [BsonElement("_id")]
    public string Id { get; set; }

    [BsonElement("casts")]
    public int Casts { get; set; }
}

File: thousands-tv/wildcard/packages/interfaces/src/db/interfaces/matchResult.ts
import { Types } from "mongoose";

export interface IMatchResults {
    _id: Types.ObjectId;
    matchResults: IMatchResult;

    __v?: number;
    createdAt?: Date;
    updatedAt?: Date;
}

export interface IMatchResults2 {
    _id: Types.ObjectId;
    matchResults: IMatchResult2;

    __v?: number;
    createdAt?: Date;
    updatedAt?: Date;
}

export interface IMatchResult {
    lobbyId: string;
    playerData: IPlayerData[];
    winningTeamId: string;
    duration: number;
    eventId: string;
    matchId: string;
    gameTypeId: string;
    createdAt: Date;
}

export interface IMatchResult2 {
    lid: string;
    pd: IPlayerData2[];
    wt: string;
    d: number;
    eid: string;
    mid: string;
    gt: string;
    createdAt: Date;
}

export interface IPlayerData {
    gamerTag: string;
    teamId: number;
    loadout: ILoadout;
    castCards: ICastCard[];
    kos: number;
    totalDamageDone: number;
    totalDamageReceived: number;
    totalHealingInflicted: number;
    totalHealingReceived: number;
    knockedouts: number;
    summonKOs: number;
    goalieKOs: number;
    sidekicksSpawned: number;
    healthCollected: string;
    manaCollected: string;
    damageTracking: IDamageTracking;
    statusTracking: IStatusTracking;
    totalWildcardsAcquired: string;
}

export interface IPlayerData2 {
    gt: string;
    t: number;
    l: ILoadout2;
    c: ICastCard2[];
    kos: number;
    tdi: number;
    tdr: number;
    thi: number;
    thr: number;
    koi: number;
    sko: number;
    gko: number;
    ss: number;
    hs: string;
    mc: string;
    dt: IDamageTracking2;
    st: IStatusTracking2;
    wca: string;
}

export interface ILoadout {
    _id: string;
    championId: string;
    name: string;
    summonCardIds: string[];
    talentCardIds: string[];
    wildCardIds: string[];
    cosmeticCardIds: string[];
    equippedCosmetics: Record<string, any>;
    IsDefault: boolean;
}

export interface ILoadout2 {
    _id: string;
    c: string;
    n: string;
    s: string[];
    t: string[];
    w: string[];
    cos: string[];
    eq: Record<string, any>;
    IsDefault: boolean;
}

export interface ICastCard {
    _id: string;
    casts: number;
}

export interface ICastCard2 {
    _id: string;
    c: number;
}

export interface IDamageTracking {
    physicalDamageInflicted: string;
    physicalDamageReceived: string;
    fireDamageInflicted: string;
    fireDamageReceived: string;
    poisonDamageInflicted: string;
    poisonDamageReceived: string;
    shockDamageInflicted: string;
    shockDamageReceived: string;
    sonicDamageInflicted: string;
    sonicDamageReceived: string;
}

export interface IDamageTracking2 {
    pdi: string;
    pdr: string;
    fdi: string;
    fdr: string;
    poisonDamageInflicted: string;
    poisonDamageReceived: string;
    kdi: string;
    kdr: string;
    sonicDamageInflicted: string;
    sonicDamageReceived: string;
}

export interface IStatusTracking {
    knockbackStatusInflicted: string;
    knockbackStatusReceived: string;
    tauntStatusInflicted: string;
    tauntStatusReceived: string;
    hasteStatusInflicted: string;
    hasteStatusReceived: string;
    slowStatusInflicted: string;
    slowStatusReceived: string;
    hackStatusInflicted: string;
    hackStatusReceived: string;
}

export interface IStatusTracking2 {
    kbi: string;
    kbr: string;
    ti: string;
    tr: string;
    hi: string;
    hr: string;
    si: string;
    sr: string;
    hai: string;
    har: string;
}